dartagnan/src/main/java/com/dat3m/dartagnan/encoding/PropertyEncoder.java

```diff
@@ -204,8 +205,32 @@ private BooleanFormula encodeLastCoConstraints() {
                     BooleanFormula sameAddress = context.sameAddress(init, w1);
                     Formula v2 = ExpressionEncoder.getLastMemValueExpr(init.getBase(), init.getOffset(), fmgr);
                     BooleanFormula sameValue = context.equal(context.value(w1), v2);
-                    enc.add(bmgr.implication(bmgr.and(lastCoExpr, sameAddress), sameValue));
+                    if (Arch.coIsTotal(program.getArch())) {
+                        enc.add(bmgr.implication(bmgr.and(lastCoExpr, sameAddress), sameValue));
+                    }
+                }
+            }
+        }
+        if (doEncodeFinalAddressValues && !Arch.coIsTotal(program.getArch())) {
+            // Coherence is not guaranteed to be total in all models (e.g., PTX),
+            // but the final value of a location should always match that of some coLast event.
+            // lastCo(w) => (lastVal(w.address) = w.val)
+            //           \/ (exists w2 : lastCo(w2) /\ lastVal(w.address) = w2.val))
+            for (Init init : program.getEvents(Init.class)) {
+                BooleanFormula lastValueEnc = bmgr.makeFalse();
+                BooleanFormula lastStoreExistsEnc = bmgr.makeFalse();
+                Formula v2 = ExpressionEncoder.getLastMemValueExpr(init.getBase(), init.getOffset(), fmgr);
+                for (Store w : program.getEvents(Store.class)) {
+                    if (!alias.mayAlias(w, init)) {
+                        continue;
+                    }
+                    BooleanFormula isLast = lastCoVar(w);
+                    BooleanFormula sameAddr = context.sameAddress(init, w);
+                    BooleanFormula sameValue = context.equal(context.value(w), v2);
+                    lastValueEnc = bmgr.or(lastValueEnc, bmgr.and(isLast, sameAddr, sameValue));
+                    lastStoreExistsEnc = bmgr.or(lastStoreExistsEnc, bmgr.and(isLast, sameAddr));
                 }
+                enc.add(bmgr.implication(lastStoreExistsEnc, lastValueEnc));
```

Comment:
As I commented before, you need to make sure that if no store gets executed, then the last value matches the initial value!

Reply:
Updated